contracts/HATVaultsRegistry.sol

```diff
@@ -0,0 +1,563 @@
+            _maxBounty,
+            _bountySplit,
+            _bountyVestingParams[0],
+            _bountyVestingParams[1]
+        );
+    }
+
+    /**
+    * @notice change the UI visibility of a vault
+    * only calleable by the owner of the contract
+    * @param _vault the vault to update
+    * @param _visible is this vault visible in the UI
+    * This parameter can be used by the UI to include or exclude the vault
+    */
+    function setVaultVisibility(address _vault, bool _visible) external onlyOwner {
+        emit SetVaultVisibility(_vault, _visible);
+    }
+
+    /**
+    * @notice change the description of a vault
+    * only calleable by the owner of the contract
+    * @param _vault the vault to update
+    * @param _descriptionHash the hash of the vault's description.
+    */
+    function setVaultDescription(address _vault, string memory _descriptionHash) external onlyOwner {
+        emit SetVaultDescription(_vault, _descriptionHash);
+    }
+
+    /**
+    * @notice Transfer the part of the bounty that is supposed to be swapped
+    * into HAT tokens from the HATVault to the registry, and keep track of the
+    * amounts to be swapped and sent/burnt in a later transaction
+    * @param _asset The vault's native token
+    * @param _hacker The address of the beneficiary of the bounty
+    * @param _hackersHatReward The amount of the vault's native token to be
+    * swapped to HAT tokens and sent to the hacker via a vesting contract
+    * @param _governanceHatReward The amount of the vault's native token to be
+    * swapped to HAT tokens and sent to governance
+    */
+    function addTokensToSwap(
+        IERC20 _asset,
+        address _hacker,
+        uint256 _hackersHatReward,
+        uint256 _governanceHatReward
+    ) external {
+        uint256 amount = _hackersHatReward + _governanceHatReward;
+        hackersHatReward[address(_asset)][_hacker] += _hackersHatReward;
+        governanceHatReward[address(_asset)] += _governanceHatReward;
+        _asset.safeTransferFrom(msg.sender, address(this), amount);
+    }
+
+    /**
+    * @notice Called by governance to swap vault's native tokens to HAT tokens
+    * and distribute the HAT tokens: Send to governance their share and send to
+    * beneficiaries their share through a vesting contract.
+    * @param _asset The vault's native token
+    * @param _beneficiaries Addresses of beneficiaries
+    * @param _amountOutMinimum Minimum amount of HAT tokens at swap
+    * @param _routingContract Routing contract to call for the swap
+    * @param _routingPayload Payload to send to the _routingContract for the
+    * swap
+    */
+    function swapAndSend(
+        address _asset,
+        address[] calldata _beneficiaries,
+        uint256 _amountOutMinimum,
+        address _routingContract,
+        bytes calldata _routingPayload
+    ) external onlyOwner {
+        // Needed to avoid a stack too deep error
+        SwapData memory swapData;
```

Comment:
so this is a bit surprising to me as the code is not so complex - you sure of ths @ben-kaufman ?

Reply:
Yes I tried any other way but it gets to stack too deep

Reply:
have you tried putting stuff in a scope? the for loop can be in {} i think

Reply:
Yes I tried any nicer trick but still got the stack too deep issue.

Reply:
you can also get rid of `totalHackerReward` and calculate the gov reward directly in the end (swapData.hatsReceived * governanceHatReward[_asset] / swapData.amount )

Reply:
that's one var, but maybe would be enough? :)

Reply:
Even without these I get stack too deep